### PR TITLE
Kraken

### PR DESCRIPTION
**Pull request by tvb**
Adding the Kraken Exchange

### CODE REVIEW

**Comment by nicka** [comment]
Sorry, please rebase... Added CI

**Comment by tvb** [comment]
@nicka done. I will fix the errors.

**Comment by nicka** [comment]
`eslint --fix` 😉

File: lib/kraken.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    this.currentPaymentMethod = null;
+    this.cryptoCurrency = process.env.PREFERRED_CRYPTO_CURRENCY || 'ETH';
+    this.cryptoCurrencies = ['XBT', 'ETH', 'LTC'];
+    this.localCurrency = process.env.PREFERRED_LOCAL_CURRENCY || 'EUR';
+    this.preferredWallet = process.env.PREFERRED_WALLET || 'fiat_account';
+    this.buyAmount = process.env.BUY_AMOUNT || 1;
+    this.sellAmount = process.env.SELL_AMOUNT || 1;
+  }
+
+  getBalance() {
```
Comment (nicka) [inline lib/kraken.js]: Could you please remove this.

File: handler.js [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
 const coinbase = new Coinbase();
+const kraken = new Kraken();
 const sms = new Sms();
 
 export const getExchangeRates = (event, context, callback) =>
   Promise.all([
     coinbase.getExchangeRates('Buy'),
     coinbase.getExchangeRates('Sell'),
+    kraken.getExchangeRates('Buy'),
+    kraken.getExchangeRates('Ask'),
```
Comment (nicka) [inline handler.js]: Would it be possible to make Kraken optional? Maybe through an ENVIRONMENT variable?

Maybe something like this?

```js
export const getExchangeRates = (event, context, callback) => {
  const promisesQueue = [
    coinbase.getExchangeRates('Buy'),
    coinbase.getExchangeRates('Sell'),
  ];

  if (process.env.KRAKEN_ENABLED === 'enabled') { // Default to disabled
    promisesQueue.push(kraken.getExchangeRates('Buy'));
    promisesQueue.push(kraken.getExchangeRates('Ask'));
  }

  return Promise.all(promisesQueue)
    .then(() =>
      callback(null, { statusCode: 200, body: { message: 'Collected exchange rates.' } }),
    ...
```

**Comment by tvb** [comment]
@nicka 

Not quite:
```
$ eslint . --fix
Cannot find module 'eslint-config-airbnb-base'
Referenced from: coinboss/.eslintrc
Error: Cannot find module 'eslint-config-airbnb-base'
Referenced from: coinboss/.eslintrc
    at ModuleResolver.resolve (/usr/local/lib/node_modules/eslint/lib/util/module-resolver.js:74:19)
    at resolve (/usr/local/lib/node_modules/eslint/lib/config/config-file.js:515:25)
    at load (/usr/local/lib/node_modules/eslint/lib/config/config-file.js:584:26)
    at configExtends.reduceRight (/usr/local/lib/node_modules/eslint/lib/config/config-file.js:421:36)
    at Array.reduceRight (native)
    at applyExtends (/usr/local/lib/node_modules/eslint/lib/config/config-file.js:405:28)
    at loadFromDisk (/usr/local/lib/node_modules/eslint/lib/config/config-file.js:556:22)
    at Object.load (/usr/local/lib/node_modules/eslint/lib/config/config-file.js:592:20)
    at Config.getLocalConfigHierarchy (/usr/local/lib/node_modules/eslint/lib/config.js:228:44)
    at Config.getConfigHierarchy (/usr/local/lib/node_modules/eslint/lib/config.js:182:43)
```

File: lib/coinbase.js [hunk 3]
```diff
@@ -22,16 +22,17 @@ export default class Coinbase {
...
-          .then((price) => {
+          .then((value) => {
             const body = {
               amount: 1,
               currency: cryptoCurrency,
               localCurrency: this.localCurrency,
-              price,
+              value,
               metricName: `${action}Price`,
+              namespace: `Coinbase`,
```
Comment (nicka) [inline lib/coinbase.js]: Please leave as is... All metrics should be within the main `coinboss` namespace. If you want to make them more specific please update `cloudWatch.js` to include something like:

```js
static setupBuySellDimensions(currency, localCurrency, provider = 'coinbase') {
    return [
      {
        Name: 'Provider',
        Value: provider,
      },
      {
        Name: 'CryptoCurrency',
        Value: currency,
      },
      {
        Name: 'LocalCurrency',
        Value: localCurrency,
      },
      {
        Name: 'Stage',
        Value: process.env.STAGE || 'dev',
      },
    ];
  }
```

**Review by nicka — CHANGES_REQUESTED** [review]

**Comment by nicka** [comment]
Did you run `yarn` before running `eslint`? 😊

**Comment by tvb** [comment]
I did now, same outcome.

**Comment by lukasnagl** [comment]
Likely `yarn eslint . --fix` if you haven’t installed `eslint` globally.

**Comment by tvb** [comment]
I ran `npm i -g eslint` but `yarn eslint . --fix` seems to work without errors. Thanks @j4zz

**Comment by lukasnagl** [comment]
@nicka Check out [eslint-staged](https://github.com/okonet/lint-staged) with husky if you don’t know it yet. Likely makes it easier for contributors to follow project-style since you could, for example, let the project automatically run formatting steps on a `precommit`.

File: handler.js (hunk 2 above)
Comment (tvb) [reply handler.js]: Sure, seems logical.

File: lib/coinbase.js (hunk 3 above)
Comment (tvb) [reply lib/coinbase.js]: Adding provider seems ok.  But what is the main namespace being called now? `Coinbase`, `coinbase` or `coinboss`.  I see three different namespaces now.

File: lib/kraken.js (hunk 1 above)
Comment (tvb) [reply lib/kraken.js]: Sure but why?

File: lib/coinbase.js (hunk 3 above)
Comment (nicka) [reply lib/coinbase.js]: If you look at the CloudFormation alarms the namespace is always `Namespace: ${self:service}` which is coming from `service: ${file(package.json):name}`. Resulting in `coinboss`.

Comment (tvb) [reply lib/coinbase.js]: Exactly, and you are mentioning `coinbase` as main namespace. So I got confused.

File: lib/kraken.js (hunk 1 above)
Comment (nicka) [reply lib/kraken.js]: Think it should be inside https://github.com/nicka/coinboss/pull/3 if we're going to merge it. For now, I think the balance does not add that much value.

**Review by nicka — CHANGES_REQUESTED** [review]

File: lib/kraken.js (hunk 1 above)
Comment (tvb) [reply lib/kraken.js]: Ok I will remove it.

**Comment by tvb** [comment]
got 2 ✅  🎉 !

**Comment by nicka** [comment]
Hehe for now! I'm adding some more Jest config tomorrow since the coverage did not check for the new kraken.js

**Comment by nicka** [comment]
Also please update the .env.example and README.md for Kraken support.

**Comment by tvb** [comment]
@nicka done

**Comment by nicka** [comment]
Build is failing

**Comment by lukasnagl** [comment]
@tvb What’s the status on this? I do think Kraken support would be valuable.

**Comment by tvb** [comment]
@j4zz I moved on, but feel free to fix this PR and have @nicka include it 👍

**Pull request closed without merging**